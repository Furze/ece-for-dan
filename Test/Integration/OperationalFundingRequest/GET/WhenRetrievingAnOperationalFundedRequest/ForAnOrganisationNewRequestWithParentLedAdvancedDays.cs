using System;
using System.Collections.Generic;
using System.Linq;
using Bard;
using MoE.ECE.CLI.Data;
using MoE.ECE.Domain.Read.Model.OperationalFunding;
using MoE.ECE.Integration.Tests.Chapter;
using MoE.ECE.Integration.Tests.Infrastructure;
using Shouldly;
using Xunit;
using Xunit.Abstractions;

namespace MoE.ECE.Integration.Tests.OperationalFundingRequest.GET.WhenRetrievingAnOperationalFundedRequest
{
    public class ForAnOrganisationNewRequestWithParentLedAdvancedDays : IntegrationTestBase<ECEStoryBook, ECEStoryData>
    {
        private Guid _businessEntityId;

        public ForAnOrganisationNewRequestWithParentLedAdvancedDays(RunOnceBeforeAllTests testSetUp,
            ITestOutputHelper output, TestState<ECEStoryBook, ECEStoryData> testState) : base(testSetUp, output,
            testState)
        {
        }

        protected override void Arrange() =>
            Given
                .A_rs7_has_been_created(rs7 =>
                {
                    rs7.OrganisationId = ReferenceData.EceServices.LeestonPlaycentre.RefOrganisationId;
                })
                .An_rs7_is_ready_for_internal_ministry_review(rs7 =>
                {
                    if (rs7.AdvanceMonths != null)
                    {
                        foreach (var advanceMonth in rs7.AdvanceMonths)
                        {
                            advanceMonth.Sessional = null;
                            advanceMonth.AllDay = null;
                            advanceMonth.ParentLed = 5;
                        }
                    }
                })
                .GetResult(data => _businessEntityId = data.Rs7Model.BusinessEntityId.GetValueOrDefault());

        protected override void Act() =>
            When.Get($"api/operational-funding-requests?business-entity-id={_businessEntityId}&revision-number=1");

        [Fact]
        public void ThenTheResponseShouldBeOk() => Then.Response.ShouldBe.Ok();

        [Fact]
        public void ThenTheResponseShouldContain1Item()
        {
            var result = Then.Response.Content<ICollection<OperationalFundingRequestModel>>();

            result.ShouldNotBeNull();
            result.Count.ShouldBe(1);
            result.ElementAt(0).AdvanceMonths.ShouldNotBeNull();
            result.ElementAt(0).AdvanceMonths?.ElementAt(0).ParentLedWorkingDays.ShouldNotBeNull();
            result.ElementAt(0).AdvanceMonths?.ElementAt(0).ParentLedWorkingDays.ShouldBe(5);
        }

        [Fact]
        public void Then_the_response_snapshot_should_be_ok() =>
            Then.Snapshot()
                .Match<ICollection<OperationalFundingRequestModel>>(IgnoreFieldsFor
                    .CollectionOperationalFundingRequestModel);
    }
}