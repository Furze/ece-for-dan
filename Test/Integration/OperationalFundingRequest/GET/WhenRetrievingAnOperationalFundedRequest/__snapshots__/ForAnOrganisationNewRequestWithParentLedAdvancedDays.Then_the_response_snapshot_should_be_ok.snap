[
  {
    "EntitlementMonths": [
      {
        "MonthName": "February 2020",
        "MonthNumber": 2,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 12,
        "SessionalNonCertificatedTeacherHours": 24,
        "TotalWorkingDays": 12,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-02-03T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 216,
            "OperatingDays": 12
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-02-03T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 12,
            "OperatingDays": 12
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-02-03T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 276,
            "OperatingDays": 12
          }
        ]
      },
      {
        "MonthName": "March 2020",
        "MonthNumber": 3,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 13,
        "SessionalNonCertificatedTeacherHours": 26,
        "TotalWorkingDays": 13,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-03-02T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 234,
            "OperatingDays": 13
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-03-02T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 13,
            "OperatingDays": 13
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-03-02T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 299,
            "OperatingDays": 13
          }
        ]
      },
      {
        "MonthName": "April 2020",
        "MonthNumber": 4,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 13,
        "SessionalNonCertificatedTeacherHours": 26,
        "TotalWorkingDays": 13,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-04-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 234,
            "OperatingDays": 13
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-04-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 13,
            "OperatingDays": 13
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-04-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 299,
            "OperatingDays": 13
          }
        ]
      },
      {
        "MonthName": "May 2020",
        "MonthNumber": 5,
        "Year": 2020,
        "AllDayCertificatedTeacherHours": 0,
        "AllDayNonCertificatedTeacherHours": 0,
        "SessionalCertificatedTeacherHours": 13,
        "SessionalNonCertificatedTeacherHours": 26,
        "TotalWorkingDays": 13,
        "WashUpPlusTen": 0.0,
        "WashUpUnderTwo": 0.0,
        "WashUpTwoAndOver": 0.0,
        "WashUpTwentyHours": 0.0,
        "TotalEntitlement": 0.0,
        "TotalFundsAdvanced": 0.0,
        "TotalEntitlementUnderTwo": 0.0,
        "TotalEntitlementTwoAndOver": 0.0,
        "TotalEntitlementTwentyHours": 0.0,
        "TotalEntitlementPlusTen": 0.0,
        "TotalWashUp": 0.0,
        "FundingComponents": [
          {
            "SessionType": "Sessional",
            "FundingComponentType": "UnderTwo",
            "StartDate": "2020-05-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 234,
            "OperatingDays": 13
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "TwoAndOver",
            "StartDate": "2020-05-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 13,
            "OperatingDays": 13
          },
          {
            "SessionType": "Sessional",
            "FundingComponentType": "PlusTen",
            "StartDate": "2020-05-01T00:00:00",
            "Amount": 0.0,
            "Rate": 0.0,
            "RateName": "",
            "FundedChildHours": 299,
            "OperatingDays": 13
          }
        ]
      }
    ],
    "OrganisationId": 5297,
    "BusinessEntityId": "fb97f7b2-99ae-4326-9c27-624ded82e0ca",
    "RequestId": "102004",
    "FundingPeriodMonth": "July",
    "FundingYear": 2021,
    "RevisionNumber": 1,
    "TotalWashUp": 0.0,
    "TotalAdvance": 0.0,
    "AdvanceMonths": [
      {
        "MonthName": "July 2020",
        "MonthNumber": 7,
        "Year": 2020,
        "ParentLedWorkingDays": 5,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 5,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "August 2020",
        "MonthNumber": 8,
        "Year": 2020,
        "ParentLedWorkingDays": 5,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 5,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "September 2020",
        "MonthNumber": 9,
        "Year": 2020,
        "ParentLedWorkingDays": 5,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 5,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      },
      {
        "MonthName": "October 2020",
        "MonthNumber": 10,
        "Year": 2020,
        "ParentLedWorkingDays": 5,
        "AllDayWorkingDays": null,
        "SessionalWorkingDays": 0,
        "AmountPayablePlusTen": 0.0,
        "AmountPayableUnderTwo": 0.0,
        "AmountPayableTwoAndOver": 0.0,
        "AmountPayableTwentyHours": 0.0,
        "TotalDays": 5,
        "TotalAdvance": 0.0,
        "FundingComponents": []
      }
    ],
    "MatchingAdvanceMonths": [],
    "Equity": null
  }
]
